### PR TITLE
refactor: Component rebalancing is now performed via a global queue

### LINKED ISSUE #2263: RouterComponent.pushOverlay() causes concurrent modification exception.
# Current bug behavior
In the update method of a component in the component tree, I am detecting the game over condition and calling the router's pushOverlay() method to push a game over dialog.  When pushOverlay is called, the remainder of the component tree finishes the update methods for the current game loop tick as is normal, but the router is going ahead with adding the overlay route to the component tree, resulting in a concurrent modification exception.

# Expected behavior
Expected behavior is to be able to push the route, allow the game tick to complete and then the overlay route be rendered on the screen.

# Steps to reproduce
1 - call the router component's pushOverlay method from a component in the tree, such as below:
```dart
@override
  void update(double dt) {
    debugger(when: gameStatus == GameStatus.gameOver);
    if (gameStatus == GameStatus.gameOver) {
      print('game is over');
      if (!_showingOverlay) {
        _showingOverlay = true;
        gameRef.router.pushOverlay('gameOver');
      }
      return;
    }
```

# More environment information

# Log information

```
======== Exception caught by scheduler library =====================================================
The following ConcurrentModificationError was thrown during a scheduler callback:
Concurrent modification during iteration: Instance of 'SplayTreeSet<LinkedHashSet<Component>>'.

When the exception was thrown, this was the stack: 
#0      _SplayTreeIterator.moveNext (dart:collection/splay_tree.dart:763:7)
#1      ExpandIterator.moveNext (dart:_internal/iterable.dart:477:21)
#2      IterableMixin.forEach (dart:collection/iterable.dart:45:23)
#3      Component.updateTree (package:flame/src/components/core/component.dart:479:16)
#4      Component.updateTree.<anonymous closure> (package:flame/src/components/core/component.dart:479:33)
```

# More information

I have not determined a cause or fix yet, but I will be happy to assist with a PR for this issue.

### PR DESCRIPTION
**Pull request by st-pasha**
# Description

This PR ensures that all component rebalancing operations are resolved from a single location, after the `update` stage but before the `render` stage (thus, components may get reordered during the update, and these changes will go into effect during the rendering step on the same game tick).

This also fixes the problem where the child changing the priorities of its parent would cause a ConcurrentModificationError.

A number of methods that were used to handle rebalancing are now marked as deprecated. From the user's perspective, the only API they should be using is the `.priority` setter.

## Checklist

- [x] I have followed the [Contributor Guide] when preparing my PR.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

- [-] Yes, this PR is a breaking change.
- [x] No, this PR is not a breaking change.

## Related Issues

Closes #2263

[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org
[CHANGELOG]: https://github.com/flame-engine/flame/blob/main/CHANGELOG.md

### CODE REVIEW

File: packages/flame/lib/src/components/core/component.dart [hunk 1]
```diff
@@ -728,22 +727,26 @@ class Component {
...
+      if (game != null && _parent != null) {
+        (game as FlameGame).enqueueRebalance(_parent!);
+      }
     }
   }
 
   /// Usually this is not something that the user would want to call since the
   /// component list isn't re-ordered when it is called.
   /// See FlameGame.changePriority instead.
+  @Deprecated('Will be removed in 1.8.0. Use priority setter instead.')
```
Comment (spydon) [inline packages/flame/lib/src/components/core/component.dart:742]: 🥳
Nice, so much cleaner without this!

**Review by spydon — APPROVED** [review]
Good stuff! This should make it quite easy to introduce global priorities too right?

File: packages/flame/test/components/priority_test.dart [hunk 2]
```diff
@@ -1,27 +1,11 @@
...
-    super.reorderChildren();
-  }
-
-  void assertCalled(int n) {
-    expect(callCount, n);
-    callCount = 0;
-  }
-}
+import '../custom_component.dart';
+// ignore_for_file: deprecated_member_use_from_same_package
```
Comment (luanpotter) [inline packages/flame/test/components/priority_test.dart:8]: what does it use that is deprecated?

**Review by luanpotter — APPROVED** [review]
LGTM with one comment

File: packages/flame/test/components/priority_test.dart (hunk 2 above)
Comment (st-pasha) [reply packages/flame/test/components/priority_test.dart:8]: I believe it's mostly calls like `game.children.changePriority(firstComponent, 11);`. Some time later we will refactor the test file to use more straightforward `firstComponent.priority = 11;`, but for now I wanted to leave everything mostly as-is to ensure that the change is non-breaking.

**Pull request merged**